junit5-extension/src/test/java/io/kroxylicious/testing/kafka/junit5ext/TemplateTest.java

```diff
@@ -62,40 +65,48 @@ public void testMultipleClusterSizesWithAdminParameters(@DimensionMethodSource(v
         assertEquals(admin.describeCluster().nodes().get().size(), cluster.getNumOfBrokers());
     }
 
-    static Stream<BrokerConfig> compression() {
-        return Stream.of(
-                brokerConfig("compression.type", "zstd"),
-                brokerConfig("compression.type", "snappy"));
-    }
-
     static Set<List<Object>> observedCartesianProduct = new HashSet<>();
 
     @Nested
     @TestInstance(TestInstance.Lifecycle.PER_CLASS)
     public class CartesianProduct {
+        static Stream<BrokerConfig> compression() {
+            return Stream.of(
+                    brokerConfig("compression.type", "zstd"),
+                    brokerConfig("compression.type", "snappy"));
+        }
+
         @TestTemplate
-        public void testCartesianProduct(@DimensionMethodSource(value = "clusterSizes", clazz = TemplateTest.class) @DimensionMethodSource(value = "compression", clazz = TemplateTest.class) KafkaCluster cluster,
+        public void testCartesianProduct(@DimensionMethodSource(value = "clusterSizes", clazz = TemplateTest.class) @DimensionMethodSource(value = "compression") KafkaCluster cluster,
                                          Admin admin)
                 throws ExecutionException, InterruptedException {
+            // Given
             assertSameCluster(cluster, admin);
-            int numBrokers = admin.describeCluster().nodes().get().size();
+            Utils.awaitExpectedBrokerCountFromBootstrapServers(cluster.getKafkaClientConfiguration(), 30, TimeUnit.SECONDS, cluster.getNumOfBrokers());
+
             ConfigResource resource = new ConfigResource(ConfigResource.Type.BROKER, "0");
+
+            // When
             Config configs = admin.describeConfigs(List.of(resource)).all().get().get(resource);
-            var compression = configs.get("compression.type").value();
 
+            // Then
+            int actualClusterSize = admin.describeCluster().nodes().get().size();
+            assertThat(actualClusterSize).as("Expected cluster to have %s nodes", cluster.getNumOfBrokers()).isEqualTo(cluster.getNumOfBrokers());
```

Comment:
Gives a clearer error message when the cluster doesn't have the expected size. Which I think leads to clearer errors when the test fails.